cirq/protocols/equal_up_to_global_phase.py

```diff
@@ -0,0 +1,171 @@
+# limitations under the License.
+
+from collections import Iterable
+import cirq
+
+
+from typing import Any, Union
+
+import numpy as np
+
+from typing_extensions import Protocol
+
+
+class SupportsEqualUpToGlobalPhase(Protocol):
+    """Object which can be compared for equality mod global phase."""
+
+    def _equal_up_to_global_phase_(
+            self,
+            other: Any,
+            *,
+            atol: Union[int, float]
+        ) -> bool:
+        """Approximate comparator.
+
+        Types implementing this protocol define their own logic for comparison
+        with other types.
+
+        Args:
+            other: Target object for comparison of equality up to global phase.
+            atol: The minimum absolute tolerance. See np.isclose() documentation
+                  for details.
+
+        Returns:
+            True if objects are equal up to a global phase, False otherwise.
+            Returns NotImplemented when checking equality up to a global phase
+            is not implemented for given types.
+        """
+
+
+def equal_up_to_global_phase(
+    val: Any,
+    other: Any,
+    *,
+    atol: Union[int, float] = 1e-8
+) -> bool:
+    """Determine whether two objects are equal up to global phase.
+
+    If `val` implements SupportsEqualUpToGlobalPhase protocol then it is
+    invoked and takes precedence over all other checks:
+     - For complex primitive type the magnitudes of the values are compared.
+     - For `val` and `other` both iterable of the same length, consecutive
+       elements are compared recursively. Types of `val` and `other` does not
+       necessarily needs to match each other. They just need to be iterable and
+       have the same structure.
+     - For all other types, fall back to _approx_eq_
+
+    Args:
+        val: Source object for approximate comparison.
+        other: Target object for approximate comparison.
+        atol: The minimum absolute tolerance. This places an upper bound on
+        the differences in _magnitudes_ of two compared complex numbers.
+
+    Returns:
+        True if objects are approximately equal up to phase, False otherwise.
+    """
+
+    # fall back to _equal_up_to_global_phase_ for val.
+    eq_up_to_phase_getter = getattr(val, '_equal_up_to_global_phase_', None)
+    if eq_up_to_phase_getter is not None:
+        result = eq_up_to_phase_getter(other, atol)
+        if result is not NotImplemented:
+            return result
+
+    # fall back to _equal_up_to_global_phase_ for other.
+    other_eq_up_to_phase_getter = getattr(
+        other, '_equal_up_to_global_phase_', None
+    )
+    if other_eq_up_to_phase_getter is not None:
+        result = other_eq_up_to_phase_getter(val, atol)
+        if result is not NotImplemented:
+            return result
+
+    # Try to compare the magnitude of two complex numbers.
+    # FIXME: why restrict casting int, float -> complex for comparison?

```

Comment:
I agree. Merge them all together into a common thing where their `abs` is compared. (Use the builtin `abs` instead of `np.abs`).